components/eam/src/physics/cam/hetfrz_classnuc.F90

```diff
@@ -310,272 +250,467 @@ subroutine hetfrz_classnuc_calc( &
+                                              r_bc, r_dust_a1, r_dust_a3, &                           ! in
+                                              do_bc, do_dst1, do_dst3, &                              ! in
+                                              frzbcimm, frzduimm)                                     ! out
+
+
+   !!! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   !   Deposition nucleation
+   !!! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+
+   ! critical germ size
+   ! assume 98% RH in mixed-phase clouds (Korolev & Isaac, JAS 2006)
+   rgdep=2*vwice*sigma_iv/(kboltz*temperature*log(rhwincloud*supersatice)) 
+
+   call calculate_hetfrz_deposition_nucleation(deltat, temperature, uncoated_aer_num, & ! in
+                                               sigma_iv, eswtr, vwice, rgdep, &         ! in
+                                               r_bc, r_dust_a1, r_dust_a3, &            ! in
+                                               do_bc, do_dst1, do_dst3, &               ! in
+                                               frzbcdep, frzdudep)                      ! out
+   
     
+
+   !!! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   ! contact nucleation
+   !!! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+   call calculate_hetfrz_contact_nucleation(deltat, temperature, uncoated_aer_num, icnlx, & ! in
+                                            sigma_iv, eswtr, rgimm, &                       ! in
+                                            r_bc, r_dust_a1, r_dust_a3, &                   ! in
+                                            Kcoll_bc, Kcoll_dust_a1, Kcoll_dust_a3, &       ! in
+                                            do_bc, do_dst1, do_dst3, &                      ! in
+                                            frzbccnt, frzducnt, errstring)                  ! out
+
+end subroutine  hetfrz_classnuc_calc
+
+
+subroutine calculate_vars_for_pdf_imm(dim_theta, pdf_imm_theta)
+
+   !****************************************************************************
+   ! calculate log normal pdf which represents the occurence probability of 
+   ! one contact anlge for one particle
+   !****************************************************************************
+
+   ! output
+   real(r8), intent(out) :: dim_theta(pdf_n_theta)
+   real(r8), intent(out) :: pdf_imm_theta(pdf_n_theta)
+
+   ! local variables
+   real(r8),parameter :: theta_min = 1._r8/180._r8*pi
+   real(r8),parameter :: theta_max = 179._r8/180._r8*pi
+   real(r8) :: x1_imm
+   real(r8) :: x2_imm
+   real(r8) :: norm_theta_imm
+   real(r8),parameter :: imm_dust_mean_theta = 46.0_r8/180.0_r8*pi
+   real(r8),parameter :: imm_dust_var_theta = 0.01_r8
+   integer :: ibin
+
+   ! calculate the integral in the denominator
+   x1_imm = (log(theta_min)-log(imm_dust_mean_theta))/(sqrt(2.0_r8)*imm_dust_var_theta)
+   x2_imm = (log(theta_max)-log(imm_dust_mean_theta))/(sqrt(2.0_r8)*imm_dust_var_theta)
```

Comment:
This subroutine should only executed once for future C++ version instead of repeating every time step, so it won't cost much. I defined some variables to store the log values to avoid repeated calculations.